hardware/is_camera/src/camera.cc

```diff
@@ -288,18 +294,43 @@ namespace is_camera {
       exit(EXIT_FAILURE);
     }
 
-    if (bayer_enable_) {
-      if (!camera.GetUInt("bayer_throttle_ratio", &bayer_throttle_ratio_)) {
-        FF_FATAL("Bayer throttle ratio not specified.");
-        exit(EXIT_FAILURE);
-      }
+    if (!camera.GetUInt("bayer_throttle_ratio", &bayer_throttle_ratio_)) {
+      FF_FATAL("Bayer throttle ratio not specified.");
+      exit(EXIT_FAILURE);
     }
 
     if (thread_running_) {
       v4l_->SetParameters(camera_gain_, camera_exposure_);
     }
   }
 
+  // Enable or disable the feature timer
+  bool CameraNodelet::EnableBayerService(ff_msgs::SetBool::Request & req,
+                     ff_msgs::SetBool::Response & res) {
+    if (req.enable && !bayer_enable_) {
+      bayer_camera_topic_ = camera_topic_ + TOPIC_HARDWARE_CAM_SUFFIX_BAYER_RAW;
+      bayer_pub_ = nh_->advertise<sensor_msgs::Image>(bayer_camera_topic_, 1);
+      // Allocate space for our Bayer output msg buffer
+      for (size_t i = 0; i < kBayerImageMsgBufferLength; i++) {
+        bayer_img_msg_buffer_[i].reset(new sensor_msgs::Image());
+        bayer_img_msg_buffer_[i]->width = kImageWidth;
+        bayer_img_msg_buffer_[i]->height = kImageHeight;
+
+        // This was tested in the lab using a color test picture
+        // Images in https://github.com/nasa/astrobee/issues/434
+        bayer_img_msg_buffer_[i]->encoding = "bayer_grbg8";
+        bayer_img_msg_buffer_[i]->step = kImageWidth;
+        bayer_img_msg_buffer_[i]->data.resize(kImageWidth * kImageHeight);
+      }
+    } else if (!req.enable && bayer_enable_) {
+      bayer_pub_.shutdown();
```

Comment:
It would be cleaner to call `reset()` on the buffers here as well, to free up some memory. This version is not a memory leak exactly, but it will have some strange behavior if you enable/disable/enable.

Reply:
Ouch. I take it back! I forgot we are using a dangerous shared memory setup and downstream nodelets could still be using the memory in these buffers.

Your code is correct to keep these buffers allocated after disable, but there should also be a guard that prevents them from being reinitialized at the next enable call.

Reply:
Are you worried that we'll disable and then enable it back in less that 1s?

Reply:
Here's my perspective: We are already accepting the risk of using an unsafe memory construct, only because it has a significant performance impact in a place where we need it. I don't see a need to add new potential failure modes on top of that. I think it's safest to (1) leave the buffers in place on a disable call, (2) put a guard on the enable call initialization that checks whether the first buffer is non-null, and if so just skips the buffer initialization loop (to avoid invalidating the existing buffers).

And as always, whenever we do something that looks strange in the code, it would be good to comment explaining why.

By the way, to directly answer your question: I agree that it's unlikely somebody will disable/enable within 1 second, but generally speaking, it's more reliable to use code correctness guarantees that don't depend on assumptions about the speed of things happening. There's no real guarantee about how quickly somebody will send those commands, nor any real guarantee about how long the downstream nodelets will take to process the buffers, and that's especially true if you think about the possibility of future code changes to those nodelets by developers who might not be aware that their image buffer is a time bomb and they have to finish processing it before it explodes. (Explosion = hard-to-debug crash caused by memory corruption.)